backend/execution_cache/src/admission.go

```diff
@@ -0,0 +1,175 @@
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type OperationType string
+
+const (
+	OperationTypeAdd OperationType = "add"
+)
+
+// patchOperation is an operation of a JSON patch, see https://tools.ietf.org/html/rfc6902 .
+type patchOperation struct {
+	Op    OperationType `json:"op"`
+	Path  string        `json:"path"`
+	Value interface{}   `json:"value,omitempty"`
+}
+
+// admitFunc is a callback for admission controller logic. Given an AdmissionRequest, it returns the sequence of patch
+// operations to be applied in case of success, or the error that will be shown when the operation is rejected.
+type admitFunc func(*v1beta1.AdmissionRequest) ([]patchOperation, error)
+
+const (
+	ContentType     string = "Content-Type"
+	JsonContentType string = "application/json"
+)
+
+var (
+	universalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+// isKubeNamespace checks if the given namespace is a Kubernetes-owned namespace.
+func isKubeNamespace(ns string) bool {
+	return ns == metav1.NamespacePublic || ns == metav1.NamespaceSystem
+}
+
+// doServeAdmitFunc parses the HTTP request for an admission controller webhook, and -- in case of a well-formed
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, fmt.Errorf("Invalid method %q, only POST requests are allowed", r.Method)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("Could not read request body: %v", err)
+	}
+
+	if contentType := r.Header.Get(ContentType); contentType != JsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("Unsupported content type %q, only %q is supported", contentType, JsonContentType)
+	}
+
+	// Step 2: Parse the AdmissionReview request.
+
+	var admissionReviewReq v1beta1.AdmissionReview
+
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
```

Comment:
To me this conditioning reads:
1. if cannot deserialize request, fail and return;
2. if the review request is nil, fail and return

So it seems like we can do:
```go
_, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq)
if err != nil {
    // fail and return
}
if admissionReviewReq.Request == nil {
    w.WriteHeader(http.StatusBadRequest)
     // fail and return
}
```